fecgov/fec-cms
# 4384 Wagtail upgrade (plug Django and testing packages)

## Summary

- Resolves #4384 
- Resolves #4363
- Resolves #4426 

With a goal of upgrading Wagtail to the latest LTS, many other upgrades were required, too.

Buckle up!

### Required reviewers

- Front-ends to make sure site builds and is still secure
- Back-end to double check performance and security, particularly around the Whitenoise changes
- Content will get another ticket to QA on dev

## Impacted areas of the application

**All of it!** Instead of listing the changes here, I've added comments to the code to explain what and why.

## Screenshots

No visible changes

## Related PRs

None

## How to test

**MAKE A BACKUP OF YOUR LOCAL CMS DATABASE**

1. Get (or make) a database dump (documented elsewhere, or check shared files)
1. Upgrade Python to 3.7.9 or 3.7.10 (likely making a new virtual env)
1. Upgrade pip to 21.0.1 `pip install --upgrade pip`
1. `env` to make sure `LDFLAGS` and `CPPFLAGS` are set
    - `export LDFLAGS="-L/usr/local/opt/openssl/lib"`
    - `export CPPFLAGS="-I/usr/local/opt/openssl/include"`
1. Install the Python requirements `pip install -r requirements.txt`
1. Drop your current database (If you haven't made a backup, do it now!) `dropdb cfdm_cms_test`
1. Create a new, empty database `createdb cfdm_cms_test`
1. Import the dump into the new database `/Library/PostgreSQL/11/bin/pg_restore --dbname cfdm_cms_test --no-acl --no-owner <dump_file_path_and_name>`
1. Switch to the fec directory
1. Do these migrations: (copy and paste this block of code into your terminal)
```bash
./manage.py migrate wagtailadmin
./manage.py migrate wagtailcore
./manage.py migrate wagtaildocs
./manage.py migrate wagtailembeds
./manage.py migrate wagtailforms
./manage.py migrate wagtailimages
./manage.py migrate wagtailredirects
./manage.py migrate wagtailsearch
./manage.py migrate wagtailsearchpromotions
./manage.py migrate wagtailusers
```
11. Do the home migration `./manage.py migrate`
1. `./manage.py showmigrations` to make sure they're all checked
1. `./manage.py makemigrations` to make sure no more are necessary
1. `./manage.py migrate` in case there are any still hanging around
1. Re-create your local superuser account `./manage.py createsuperuser`
1. Install the latest npm packages, just in case `npm i`
1. Build the JavaScript and css, just in case `npm run build`
1. `pytest` to make sure all tests pass. (`pytest --maxfail=1` will only test until it hits a single failure and then stop—saves a _ton_ of logs to search!)
1. For some reason, I had to set `FEC_API_URL` because I was getting connection errors.
1. `npm run test-single` to make sure JavaScript tests pass
1. `./manage.py runserver`
1. Check many types of pages to make sure things look and work correctly.
    - homepage
    - data pages
    - legal
    - blog
    - calendar
    - others?
1. Log in to site admin to make sure it works as it should
1. Tricky: Here's an error I ran into:
   1. Log into admin
   1. Stop your `runserver` process (ctrl-c?)
   1. `./manage.py runserver`
   1. Loaded _any_ page on the site
   1. Go back to admin
   1. All good? Yay! (I was getting an error about `insufficient padding` of all things—something about 64-bit and the cookie session IDs.)
